File: 820-find-eventual-safe-states/find-eventual-safe-states.py
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        V = len(graph)
        indegree = [0] * V
        adj = [[] for _ in range(V)]

        for i in range(V):
            for j in graph[i]:
                adj[j].append(i)
                indegree[i]+=1

        # for i in range(V):
        #     for j in graph[i]:
        #         indegree[j]+=1

        q = deque()

        for i in range(V):
            if indegree[i] == 0:
                q.append(i)

        topo = []

        while len(q):
            node = q.popleft()
            topo.append(node)

            for i in adj[node]:
                indegree[i]-=1
                if indegree[i] == 0:
                    q.append(i)
                    
        topo.sort()
        return topo

File: 820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();

        vector<vector<int>> adj(n);

        vector<int> indeg(n);

        for(int i = 0; i<n; i++){
            for(auto x: graph[i]){
                adj[x].push_back(i);
                indeg[i]++;
            }
        }

        queue<int> q;

        vector<int> sf(n, 0);

        for(int i = 0; i<n; i++){
            if(indeg[i] == 0){
                q.push(i);
            }
        }

        vector<int> topo;

        while(!q.empty()){
            int node = q.front();
            q.pop();

            sf[node] = 1;

            for(auto x: adj[node]){
                indeg[x]--;
                if(indeg[x] == 0) q.push(x);
            }
        }

        for(int i = 0; i<n; i++){
            if(sf[i]) topo.push_back(i);
        }

        return topo;
    }
};